### PR TITLE
[ide] fix unconditional goto-point on editing an error (fix #9488)

### LINKED ISSUE #4988: Name discrepancy in coq-inferior.el
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4988
From: @brabalan
Reported version: 8.5
CC:   @ppedrot

### LINKED ISSUE #9488: End of input in CoqIDE
#### Version

Coq 8.8.2

#### Operating system

Red Hat Enterprise Linux Workstation 7.4

#### Description of the problem

The following actions cause an anomaly.

* Open CoqIDE.
* Insert the following code into the buffer.
* Try to navigate to the end.
* Navigate backward.
* Navigate backward.
* Delete everything from any point within `idtac 1` to the end of the document.

```
Goal True.
Proof. idtac 1. idtac 2. qed.
```

```
Anomaly "Uncaught exception Stm.End_of_input."
```

The only way to recover is to delete the remaining proof.

#### Notes

This reliably reproduces #8647.
Even though this bug may already be fixed,
this report should at least give peace of mind
to the past victims of this elusive anomaly.

### LINKED ISSUE #9552: `Qed` edits are ignored
In CoqIDE, type the following:

```coq
Lemma x : True.
Proof. trivial. Qed.
```

jump to the end. `Qed` is colored in grey. Observe that you can edit `Qed`, replacing it with what you want. Nothing is invalidated, even if there's some processed text after.

Fixed by #9549.

### PR DESCRIPTION
**Pull request by gares**
The code was trying to "back-to" when editing an error, and it was jumping to the sentence immediately before the error. Sometimes that sentence is not already processed, hence the result was "going forward".

Fixes #9488.
Fixes #9552.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
#4988 is some emacs related issue, did you mean another number?

**Comment by gares** [comment]
oops

**Comment by gares** [comment]
It also fixes the fact that one could edit a "gray" Qed.

File: ide/session.ml [hunk 1]
```diff
@@ -145,23 +145,29 @@ let set_buffer_handlers
...
+      else None in
     aux it it in
   let insert_cb it s = if String.length s = 0 then () else begin
     Minilib.log ("insert_cb " ^ string_of_int it#offset);
     let text_mark = add_mark it in
     let () = update_prev it in
     if it#has_tag Tags.Script.to_process then
       cancel_signal "Altering the script being processed in not implemented"
+    else if it#has_tag Tags.Script.incomplete then
+      cancel_signal "Altering the script being processed in not implemented"
```
Comment (maximedenes) [inline ide/session.ml]: Is this the right message?

File: ide/session.ml [hunk 2]
```diff
@@ -171,14 +177,18 @@ let set_buffer_handlers
...
+        cancel_signal "Altering the script being processed in not implemented"
       else if min_iter#has_tag Tags.Script.to_process then
         cancel_signal "Altering the script being processed in not implemented"
       else if min_iter#has_tag Tags.Script.processed then
         call_coq_or_cancel_action (coqops#go_to_mark (`MARK text_mark))
       else if min_iter#has_tag Tags.Script.error_bg then
-        let prev_sentence_end = backto_before_error min_iter in
-        let text_mark = add_mark prev_sentence_end in
-        call_coq_or_cancel_action (coqops#go_to_mark (`MARK text_mark))
+        match processed_sentence_just_before_error min_iter with
```
Comment (maximedenes) [inline ide/session.ml]: Can't we factorize all this logic between insert and delete?

File: ide/session.ml (hunk 1 above)
Comment (gares) [reply ide/session.ml]: yes, `incomplete` mean "incompletely processed"

File: ide/session.ml (hunk 2 above)
Comment (gares) [reply ide/session.ml]: you are welcome ;-)

**Review by ppedrot — APPROVED** [review]
I'll trust you on that one, cause I have no idea of the expected invariants.

**Pull request merged**